### ISSUE TITLE
undefined value should not be set as null into db

### ISSUE THREAD

**Issue by zhaoyao91** [jagi/meteor-astronomy#170]
By now, when some fields are defined in the schema, if it has no value, it will be set **null** into db. I think **undefined** value should not be set in the db,  and if a value is set undefined, it should be $unset in the db.

Setting null for undefined fields is not only unnecessary but also not correct, in my opinion.

**Comment by jagi** [comment]
It's correct if you think about it in terms of Model layer. And you can't unset a field for a single document because you want all data to be consistent. I'm planning to introduce migration feature so it will make even more sens in the future.

**Comment by zhaoyao91** [comment]
what I worry about is models with sparse or optional fields. won't the null fields waste spaces in db, or introduce other efficiency problem?

**Comment by jagi** [comment]
Not storing a null value would be ok as long as you have a default value for a field. So maybe it's a good idea for improvement. But in other cases it's pointless.

**Comment by talha-asad** [comment]
I think ```undefined``` should not store that field in db and ```null``` should store the default value from schema.

If a ```null``` value field doesn't have a default it should be stored as ```null``` in the db as well.

**Comment by jagi** [comment]
It will just store `null` if the `null` value is allowed. Default value is only used on a document insert or if the field value is undefined in the collection. In Astronomy 2.0 it will be possible to unset a field but you will be doing it on your own risk :P

**Comment by JulianKingman** [comment]
It would be nice if it just didn't include the field in the document for optional fields. For example, I'm getting duplicate username issues with the Meteor.users() collection because when it inserts without a username, it sets it to null.
What I expect:
````javascript
user: {
    _id: somefancyid,
    profile: {//etc...}
}
````
What I get:
````javascript
user: {
    _id: somefancyid,
    username: null,
    profile: {//etc...}
}
````
Along with an error that the username already exists when I try to create the user.

**Comment by jagi** [comment]
Ok right, in deed it's an issue. So I will have to charge that. Thanks for pointing this out!

**Comment by Andrew1431** [comment]
Did you ever look in to this jagi? I'm getting similar issues with sparse indexing. 

```MongoError: insertDocument :: caused by :: 11000 E11000 duplicate key error index: meteor.crm.members.$driversLicense_1  dup key: { : null }```

rc.5 here

**Comment by jagi** [comment]
I don't remember in which version I've fixed undefined/null values. You have to try out the newest version which is jagi:astronomy@2.0.0-rc.7

**Comment by lukejagodzinski** [comment]
Fixed in 2.0. Preparing to release, so I'm closing this issue.
